syntax = "proto3";


package fmi3_messages;

// https://developers.google.com/protocol-buffers/docs/reference/java-generated#package
option java_package = "";
option java_outer_classname = "Fmi3Messages";


// ----------------------- FMI3 ----------------------

enum Fmi3Status {
  FMI3_OK = 0;
  FMI3_WARNING=1;
  FMI3_DISCARD=2;
  FMI3_ERROR=3;
  FMI3_FATAL=4;
}


message Fmi3InstantiateModelExchange {
  string instance_name = 1;
  string instantiation_token = 2;
  string resource_path = 3;
  bool visible = 4;
  bool logging_on = 5;
}


message Fmi3InstantiateCoSimulation {
  string instance_name = 1;
  string instantiation_token = 2;
  string resource_path = 3;
  bool visible = 4;
  bool logging_on = 5;
  bool event_mode_used = 6;
  bool early_return_allowed = 7;
  repeated uint32 required_intermediate_variables = 8;
}

message Fmi3InstantiateScheduledExecution {
  string instance_name = 1;
  string instantiation_token = 2;
  string resource_path = 3;
  bool visible = 4;
  bool logging_on = 5;
}

message Fmi3DoStep {
  double current_communication_point = 1;
  double communication_step_size = 2;  
  bool no_set_fmu_state_prior_to_current_point = 3;
}

message Fmi3SetDebugLogging {
  bool logging_on = 1;
  repeated string categories = 2;
}

message Fmi3EnterInitializationMode {
  bool tolerance_defined = 1;
  optional double tolerance = 2;
  double start_time = 3;
  bool stop_time_defined = 4;
  optional double stop_time = 5;
}

message Fmi3ExitInitializationMode {}

message Fmi3EnterStepMode {}

message Fmi3EnterEventMode {}

message Fmi3FreeInstance {}

message Fmi3Terminate {}

message Fmi3Reset {}

message Fmi3SerializeFmuState {}

message Fmi3DeserializeFmuState {
    bytes state = 1;
}

message Fmi3UpdateDiscreteStates {}

message Fmi3EnterConfigurationMode {}

message Fmi3ExitConfigurationMode {}

// ----------------------- FMI3 Getters ----------------------


message Fmi3GetFloat32 {
  repeated uint32 value_references = 1;
}

message Fmi3GetFloat64 {
  repeated uint32 value_references = 1;
}

message Fmi3GetInt8 {
  repeated uint32 value_references = 1;
}

message Fmi3GetUInt8 {
  repeated uint32 value_references = 1;
}

message Fmi3GetInt16 {
  repeated uint32 value_references = 1;
}

message Fmi3GetUInt16 {
  repeated uint32 value_references = 1;
}

message Fmi3GetInt32 {
  repeated uint32 value_references = 1;
}

message Fmi3GetUInt32 {
  repeated uint32 value_references = 1;
}

message Fmi3GetInt64 {
  repeated uint32 value_references = 1;
}

message Fmi3GetUInt64 {
  repeated uint32 value_references = 1;
}

message Fmi3GetBoolean {
  repeated uint32 value_references = 1;
}

message Fmi3GetString {
  repeated uint32 value_references = 1;
}

message Fmi3GetBinary {
  repeated uint32 value_references = 1;
}

message Fmi3GetClock {
  repeated uint32 value_references = 1;
}

message Fmi3GetDirectionalDerivative {
  repeated uint32 unknowns = 1;
  repeated uint32 knowns = 2;
  repeated double seed = 3;
  repeated double sensitivity = 4;
}

message Fmi3GetAdjointDerivative {
  repeated uint32 unknowns = 1;
  repeated uint32 knowns = 2;
  repeated double seed = 3;
  repeated double sensitivity = 4;
}

message Fmi3GetOutputDerivatives {
  repeated uint32 value_references = 1;
  repeated uint32 orders = 2;
  repeated double values = 3;
}

// ----------------------- FMI3 Setters ----------------------

message Fmi3SetFloat32 {
  repeated uint32 value_references= 1;
  repeated float values = 2;
}

message Fmi3SetFloat64 {
  repeated uint32 value_references= 1;
  repeated double values = 2;
}

message Fmi3SetInt8 {
  repeated uint32 value_references= 1;
  repeated int32 values = 2;
}

message Fmi3SetUInt8 {
  repeated uint32 value_references= 1;
  repeated uint32 values = 2;
}

message Fmi3SetInt16 {
  repeated uint32 value_references= 1;
  repeated int32 values = 2;
}

message Fmi3SetUInt16 {
  repeated uint32 value_references= 1;
  repeated uint32 values = 2;
}

message Fmi3SetInt32 {
  repeated uint32 value_references= 1;
  repeated int32 values = 2;
}

message Fmi3SetUInt32 {
  repeated uint32 value_references= 1;
  repeated uint32 values = 2;
}

message Fmi3SetInt64 {
  repeated uint32 value_references= 1;
  repeated int64 values = 2;
}

message Fmi3SetUInt64 {
  repeated uint32 value_references= 1;
  repeated uint64 values = 2;
}

message Fmi3SetBoolean {
  repeated uint32 value_references = 1;
  repeated bool values = 2;
}

message Fmi3SetString {
  repeated uint32 value_references = 1;
  repeated string values = 2;
}

message Fmi3SetBinary {
  repeated uint32 value_references = 1;
  repeated uint64 value_sizes = 2;
  repeated bytes values = 3;
}

message Fmi3SetClock {
  repeated uint32 value_references = 1;
  repeated bool values = 2;
}

// ----------------------- FMI3 Return Values ----------------------

message Fmi3DoStepReturn
{
  Fmi3Status status = 1;
  bool event_handling_needed = 2;
  bool terminate_simulation = 3;
  bool early_return = 4;
  double last_successful_time = 5;
}

// For methods that do not return a status code
message Fmi3EmptyReturn {}

message Fmi3StatusReturn {
    Fmi3Status status = 1;
}

message Fmi3FreeInstanceReturn {}

message Fmi3GetFloat32Return {
  Fmi3Status status = 1;
  repeated float values = 2;
}

message Fmi3GetFloat64Return {
  Fmi3Status status = 1;
  repeated double values = 2;
}

message Fmi3GetInt8Return {
  Fmi3Status status = 1;
  repeated int32 values = 2;
}

message Fmi3GetUInt8Return {
  Fmi3Status status = 1;
  repeated uint32 values = 2;
}

message Fmi3GetInt16Return {
  Fmi3Status status = 1;
  repeated int32 values = 2;
}

message Fmi3GetUInt16Return {
  Fmi3Status status = 1;
  repeated uint32 values = 2;
}

message Fmi3GetInt32Return {
  Fmi3Status status = 1;
  repeated int32 values = 2;
}

message Fmi3GetUInt32Return {
  Fmi3Status status = 1;
  repeated uint32 values = 2;
}

message Fmi3GetInt64Return {
  Fmi3Status status = 1;
  repeated int64 values = 2;
}

message Fmi3GetUInt64Return {
  Fmi3Status status = 1;
  repeated uint64 values = 2;
}

message Fmi3GetBooleanReturn {
  Fmi3Status status = 1;
  repeated bool values = 2;
}

message Fmi3GetStringReturn {
  Fmi3Status status = 1;
  repeated string values = 2;
}

message Fmi3GetBinaryReturn {
  Fmi3Status status = 1;
  repeated bytes values = 2;
}

message Fmi3GetDirectionalDerivativeReturn {
  Fmi3Status status = 1;
  repeated double values = 2;
}

message Fmi3GetAdjointDerivativeReturn {
  Fmi3Status status = 1;
  repeated double values = 2;
}

message Fmi3GetOutputDerivativesReturn {
  Fmi3Status status = 1;
  repeated double values = 2;
}

message Fmi3SerializeFmuStateReturn {
  Fmi3Status status = 1;
  bytes state = 2;
}

message Fmi3GetClockReturn {
  Fmi3Status status = 1;
  repeated bool values = 2;
}

message Fmi3UpdateDiscreteStatesReturn {
  Fmi3Status status = 1;
  bool discrete_states_need_update = 2;
  bool terminate_simulation = 3;
  bool nominals_continuous_states_changed = 4;
  bool values_continuous_states_changed = 5;
  bool next_event_time_defined = 6;
  double next_event_time = 7;
}

// ----------------------- FMI3 Clock Related Functions -----------------------

enum Fmi3IntervalQualifier {
  FMI3_INTERVALNOTYETKNOWN = 0;
  FMI3_INTERVALUNCHANGED=1;
  FMI3_INTERVALCHANGED=2;
}

message Fmi3GetIntervalDecimal {
  repeated uint32 value_references = 1;
}

message Fmi3GetIntervalDecimalReturn {
  Fmi3Status status = 1;
  repeated double intervals = 2;
  repeated int32 qualifiers = 3;
}

message Fmi3GetIntervalFraction {
  repeated uint32 value_references = 1;
}

message Fmi3GetIntervalFractionReturn {
  Fmi3Status status = 1;
  repeated uint64 counters = 2;
  repeated uint64 resolutions = 3;
  repeated int32 qualifiers = 4;
}

message Fmi3GetShiftDecimal {
  repeated uint32 value_references = 1;
}

message Fmi3GetShiftDecimalReturn {
  Fmi3Status status = 1;
  repeated double shifts = 2;
}

message Fmi3GetShiftFraction {
  repeated uint32 value_references = 1;
}

message Fmi3GetShiftFractionReturn {
  Fmi3Status status = 1;
  repeated uint64 counters = 2;
  repeated uint64 resolutions = 3;
}

message Fmi3SetIntervalDecimal {
  repeated uint32 value_references= 1;
  repeated double intervals = 2;
}

message Fmi3SetIntervalFraction {
  repeated uint32 value_references= 1;
  repeated uint64 counters = 2;
  repeated uint64 resolutions = 3;
}

message Fmi3SetShiftDecimal {
  repeated uint32 value_references= 1;
  repeated double shifts = 2;
}

message Fmi3SetShiftFraction {
  repeated uint32 value_references= 1;
  repeated uint64 counters = 2;
  repeated uint64 resolutions = 3;
}

// ----------------------- FMI Command Wrapper ----------------------

message Fmi3Command {
  oneof command {

    // FMI3

    Fmi3InstantiateModelExchange Fmi3InstantiateModelExchange= 1;
    Fmi3InstantiateCoSimulation Fmi3InstantiateCoSimulation = 2;
    Fmi3InstantiateScheduledExecution Fmi3InstantiateScheduledExecution = 3;
    Fmi3DoStep Fmi3DoStep = 4;
    Fmi3SetDebugLogging Fmi3SetDebugLogging = 5;
    Fmi3EnterInitializationMode Fmi3EnterInitializationMode = 6;
    Fmi3ExitInitializationMode Fmi3ExitInitializationMode = 7;
    Fmi3FreeInstance Fmi3FreeInstance = 8;
    Fmi3Terminate Fmi3Terminate = 9;
    Fmi3Reset Fmi3Reset = 10;
    Fmi3GetFloat32 Fmi3GetFloat32 = 13;
    Fmi3GetFloat64 Fmi3GetFloat64 = 14;
    Fmi3GetInt8 Fmi3GetInt8 = 15;
    Fmi3GetUInt8 Fmi3GetUInt8 = 16;
    Fmi3GetInt16 Fmi3GetInt16 = 17;
    Fmi3GetUInt16 Fmi3GetUInt16 = 18;
    Fmi3GetInt32 Fmi3GetInt32 = 19;
    Fmi3GetUInt32 Fmi3GetUInt32 = 20;
    Fmi3GetInt64 Fmi3GetInt64 = 21;
    Fmi3GetUInt64 Fmi3GetUInt64 = 22;
    Fmi3GetBoolean Fmi3GetBoolean = 23;
    Fmi3GetString Fmi3GetString = 24;
    Fmi3GetBinary Fmi3GetBinary = 25;
    Fmi3GetDirectionalDerivative Fmi3GetDirectionalDerivative = 26;
    Fmi3GetAdjointDerivative Fmi3GetAdjointDerivative = 27;
    Fmi3GetOutputDerivatives Fmi3GetOutputDerivatives = 28;
    Fmi3SetFloat32 Fmi3SetFloat32 = 29;
    Fmi3SetFloat64 Fmi3SetFloat64 = 30;
    Fmi3SetInt8 Fmi3SetInt8 = 31;
    Fmi3SetUInt8 Fmi3SetUInt8 = 32;
    Fmi3SetInt16 Fmi3SetInt16 = 33;
    Fmi3SetUInt16 Fmi3SetUInt16 = 34;
    Fmi3SetInt32 Fmi3SetInt32 = 35;
    Fmi3SetUInt32 Fmi3SetUInt32 = 36;
    Fmi3SetInt64 Fmi3SetInt64 = 37;
    Fmi3SetUInt64 Fmi3SetUInt64 = 38;
    Fmi3SetBoolean Fmi3SetBoolean = 39;
    Fmi3SetString Fmi3SetString = 40;
    Fmi3SetBinary Fmi3SetBinary = 41;
    Fmi3SerializeFmuState Fmi3SerializeFmuState = 42;
    Fmi3DeserializeFmuState Fmi3DeserializeFmuState = 43;
	  Fmi3GetClock Fmi3GetClock = 44;
	  Fmi3SetClock Fmi3SetClock = 45;
	  Fmi3GetIntervalDecimal Fmi3GetIntervalDecimal = 46;
	  Fmi3EnterStepMode Fmi3EnterStepMode = 47;
	  Fmi3EnterEventMode Fmi3EnterEventMode = 48;
	  Fmi3UpdateDiscreteStates Fmi3UpdateDiscreteStates = 49;
    Fmi3EnterConfigurationMode Fmi3EnterConfigurationMode = 50;
    Fmi3ExitConfigurationMode Fmi3ExitConfigurationMode = 51;
    Fmi3GetIntervalFraction Fmi3GetIntervalFraction = 52;
    Fmi3GetShiftDecimal Fmi3GetShiftDecimal = 53;
    Fmi3GetShiftFraction Fmi3GetShiftFraction = 54;
    Fmi3SetIntervalDecimal Fmi3SetIntervalDecimal = 55;
    Fmi3SetIntervalFraction Fmi3SetIntervalFraction = 56;
    Fmi3SetShiftDecimal Fmi3SetShiftDecimal = 57;
    Fmi3SetShiftFraction Fmi3SetShiftFraction = 58;
  }
}

// ---------------------- FMI Return Wrapper ------------------------

message Fmi3Return {
  oneof return_message {
    Fmi3EmptyReturn empty = 1;
    Fmi3StatusReturn status = 2;
    Fmi3DoStepReturn do_step = 3;
    Fmi3FreeInstanceReturn free_instance = 4;
    Fmi3GetFloat32Return get_float_32 = 5;
    Fmi3GetFloat64Return get_float_64 = 6;
    Fmi3GetInt8Return get_int_8 = 7;
    Fmi3GetUInt8Return get_u_int_8 = 8;
    Fmi3GetInt16Return get_int_16 = 9;
    Fmi3GetUInt16Return get_u_int_16 = 10;
    Fmi3GetInt32Return get_int_32 = 11;
    Fmi3GetUInt32Return get_u_int_32 = 12;
    Fmi3GetInt64Return get_int_64 = 13;
    Fmi3GetUInt64Return get_u_int_64 = 14;
    Fmi3GetBooleanReturn get_boolean = 15;
    Fmi3GetStringReturn get_string = 16;
    Fmi3GetBinaryReturn get_binary = 17;
    Fmi3GetDirectionalDerivativeReturn get_directional_derivative = 18;
    Fmi3GetAdjointDerivativeReturn get_adjoint_derivative = 19;
    Fmi3GetOutputDerivativesReturn get_output_derivatives = 20;
    Fmi3SerializeFmuStateReturn serialize_fmu_state = 21;
    Fmi3GetClockReturn get_clock = 22;
    Fmi3UpdateDiscreteStatesReturn update_discrete_states = 23;
    Fmi3GetIntervalDecimalReturn get_interval_decimal = 24;
    Fmi3GetIntervalFractionReturn get_interval_fraction = 25;
    Fmi3GetShiftDecimalReturn get_shift_decimal = 26;
    Fmi3GetShiftFractionReturn get_shift_fraction = 27;
  }
}
